### ISSUE TITLE
elastic_exporter metric values for disk are difference than actual

### ISSUE THREAD

**Issue by kushalgangan** [justwatchcom/elasticsearch_exporter#325]
we are getting incorrect values for disk metrics

**Comment by sysadmind** [comment]
@kushalgangan Can you please elaborate more on your issue? I was not able to reproduce locally. What specific metrics are you speaking of? What are you comparing them to in order to know that they are inaccurate? What version of the exporter are you using? What version is your elasticsearch cluster? 

I checked the following locally and was able to verify that the output from the exporter matched what I could query from the REST API in elasticsearch (using `/_nodes/stats`).

- elasticsearch_filesystem_data_available_bytes
- elasticsearch_filesystem_data_free_bytes
- elasticsearch_filesystem_data_size_bytes
- elasticsearch_filesystem_io_stats_device_operations_count
- elasticsearch_filesystem_io_stats_device_read_operations_count
- elasticsearch_filesystem_io_stats_device_read_size_kilobytes_sum
- elasticsearch_filesystem_io_stats_device_write_operations_count
- elasticsearch_filesystem_io_stats_device_write_size_kilobytes_sum

I am running the master branch of the exporter and elasticsearch 7.6.0

**Comment by mohitrajain** [comment]
@sysadmind I am using following version of elasticsearch_exporter :
elasticsearch_exporter, version 1.2.1 (branch: HEAD, revision: 746222a8153db4ecaf99f239dcd34a187fdcb894)
  build user:       root@7c33e2dcfad4
  build date:       20210629-13:54:56
  go version:       go1.16.5
  platform:         linux/amd64
 
  with
  elasticsearch
  "version" : "6.2.3"
  "build_hash" : "c59ff00",
  "build_date" : "2018-03-13T10:06:29.741383Z",

I have also noticed some discrepancy with metric **elasticsearch_filesystem_data_size_bytes**. Values reported by metrics are a bit more than actual numbers. Haven't checked for all the size metrics but noticed it for **elasticsearch_filesystem_data_available_bytes** too.

Metric reported is:

elasticsearch_filesystem_data_size_bytes{cluster="elasticsearch",es_client_node="true",es_data_node="true",es_ingest_node="true",es_master_node="true",host="172.17.0.1",instance="127.0.0.1:6363",job="elasticsearchexporter",mount="/data (/dev/xvdf)",name="Random",path="/data/elasticsearch/nodes/0"} **2642099224576**

elasticsearch_filesystem_data_available_bytes{cluster="elasticsearch",es_client_node="true",es_data_node="true",es_ingest_node="true",es_master_node="true",host="172.17.0.1",instance="127.0.0.1:6363",job="elasticsearchexporter",mount="/data (/dev/xvdf)",name="Random",path="/data/elasticsearch/nodes/0"}	**101248729497**

when checked with **df**, number is different for same disk :
sudo df /data
Filesystem      1K-blocks       Used Available Use% Mounted on
/dev/xvdf      **2580175024** 1479962620 **992698348**  60% /data

Note: /data contains other data apart from elasticsearch which is some ~100GiB but that shouldn't affect total number.

I suspect the correct number is lost in translation of units. Let me know if more information is required from my part. What could be the reason for this discrepancy?

**Comment by chuck6** [comment]
> 

from my prometheus,  I use promql "elasticsearch_filesystem_data_free_bytes{}" ，but it returns no data, just as "Empty query result" .  how to  deploy the exporter to see the  monitor data?

**Comment by GloomyDay** [comment]
@mohitrajain I found that this is not exporter fault ,  curl -s 127.0.0.1:9200/_nodes/stats/fs returns this data for some reason

**Comment by VLZZZ** [comment]
I can share my experience of getting confused receiving "incorrect" filesystem data with managed AWS OpenSearch (Elasticsearch)

It turned out that the answer is difened in AWS Opensearch service docs ([Calculating storage requirements](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/bp-storage.html))
> OpenSearch Service reserves 20% of the storage space of each instance (up to 20 GiB) for segment merges, logs, and other internal operations.

_More detailed context_:
> Because of this 20 GiB maximum, the total amount of reserved space can vary dramatically depending on the number of instances in your domain. For example, a domain might have three m6g.xlarge.search instances, each with 500 GiB of storage space, for a total of 1.46 TiB. In this case, the total reserved space is only 60 GiB. Another domain might have 10 m3.medium.search instances, each with 100 GiB of storage space, for a total of 0.98 TiB. Here, the total reserved space is 200 GiB, even though the first domain is 50% larger.

And here - [Lack of available storage space](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/handling-errors.html#troubleshooting-cluster-block)

> You might receive a ClusterBlockException error for the following reasons.
> 
> Lack of available storage space
If one or more nodes in your cluster has storage space less than the minimum value of:
> 1. 20% of available storage space, or
> 2. 20 GiB of storage space, basic write operations like adding documents and creating indexes can start to fail.


So basically we were receiving proper `_nodes/stats/fs` via `elasticsearch_filesystem_data_available_bytes` of `~26GiB`  or 20% free storage space

<img width="2439" height="551" alt="Image" src="https://github.com/user-attachments/assets/86182acf-9b62-4490-98c3-f094bfe76bc7" />

While CloudWatch Metric [FreeStorageSpace](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/managedomains-cloudwatchmetrics.html) were reporting `0`

<img width="2946" height="1258" alt="Image" src="https://github.com/user-attachments/assets/1365ca7b-8f58-40a2-b5e3-a6f6d40a28c1" />

I hope this might be helpful for anyone
